src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Win32.SafeHandles;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Runtime.InteropServices;
+
+namespace System.IO
+{
+    /* This class is used by for reading from the stdin.
+     * It is designed to read stdin in raw mode for interpreting
+     * key press events and maintain its own buffer for the same.
+     * which is then used for all the Read operations
+     */
+    internal class StdInStreamReader : StreamReader
+    {
+        char[] unprocessedBufferToBeRead; // Buffer that might have already been read from stdin but not yet processed.
+        const int bytesToBeRead = 255; // No. of bytes to be read from the stream at a time.
+        int startIndex; // First unprocessed index in the buffer;
+        int endIndex; // Last unprocessed index in the buffer;
+
+        internal StdInStreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByteOrderMarks, int bufferSize, bool leaveOpen) : base(stream, encoding, detectEncodingFromByteOrderMarks, bufferSize, leaveOpen)
+        {
+            unprocessedBufferToBeRead = new char[encoding.GetMaxCharCount(bufferSize)];
+            startIndex = endIndex = -1; // There is no unprocessed index yet.
+        }
+
+        /// <summary> Checks whether the unprocessed buffer is empty. </summary>
+        bool IsExtraBufferEmpty()
+        {
+            if (startIndex == -1 && endIndex == -1) return true;
+
+            if (startIndex > endIndex) // Everything has been processed;
+            {
+                startIndex = -1;
+                endIndex = -1;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary> Reads the buffer in the raw mode. </summary>
+        unsafe void ReadExtraBuffer()
+        {
+            Debug.Assert(IsExtraBufferEmpty());
+
+            byte[] buffer = new byte[bytesToBeRead];
+            fixed (byte* bufPtr = buffer)
+            {
+                int result;
+                Interop.CheckIo(result = Interop.Sys.ReadStdinUnbuffered((byte*)bufPtr, bytesToBeRead));
+                Debug.Assert(result <= bytesToBeRead);
+
+                // Now we need to convert the byte buffer to char buffer.
+                int charCount = CurrentEncoding.GetMaxCharCount(result);
+                Debug.Assert(charCount < unprocessedBufferToBeRead.Length);
+
+                CurrentEncoding.GetChars(buffer, 0, buffer.Length, unprocessedBufferToBeRead, ++startIndex);
+                endIndex += charCount;
```

Comment:
This doesn't look right.  GetChars returns the number of characters converted, but this is ignoring that value.  Instead it's using GetMaxCharCount, which will likely produce the wrong result.  There shouldn't be a need to call GetMaxCharCount at all, and instead these four lines can just be:
```
endIndex += CurrentEncoding.GetChars(buffer, 0, buffer.Length, unprocessedBufferToRead, ++startIndex);
```
It's probably also worth a comment explaining the math here, that startIndex and endIndex are both -1, and you need endIndex to point to (not beyond) the last character and you need startIndex to be 0. Though if it were me, I'd skip the increment on startIndex and just be explicit about it:
```
startIndex = 0;
int charCount = CurrentEncoding.GetChars(buffer, 0, buffer.Length, unprocessedBufferToBeRead, startIndex);
endIndex = charCount - 1;
```